governator-core/src/main/java/com/netflix/governator/internal/PreDestroyLifecycleActions.java

```diff
@@ -2,55 +2,103 @@
 
 import com.netflix.governator.LifecycleAction;
 
 /**
- * Special AbstractLifecycleFeature to support @PreDestroy annotation processing.
- * Note that this feature is implicit in LifecycleModule and therefore does not need 
- * to be added using the LifecycleFeature multibinding.
+ * Special AbstractLifecycleFeature to support @PreDestroy annotation processing and
+ * java.io.AutoCloseable detection. Note that this feature is implicit in LifecycleModule 
+ * and therefore does not need to be added using the LifecycleFeature multibinding.
  * 
  * @author elandau
  */
-public final class PreDestroyLifecycleActions extends AbstractLifecycleFeature {
+public final class PreDestroyLifecycleActions extends AbstractLifecycleFeature<Set<String>> {
 
     public static PreDestroyLifecycleActions INSTANCE = new PreDestroyLifecycleActions();
     
     private PreDestroyLifecycleActions() {
+        super(IntrospectionFlag.METHOD, IntrospectionFlag.SUPERCLASS);
     }
-
+    
     @Override
-    protected List<LifecycleAction> getMethodActions(final Class<?> type, final Method method) {
-        if (null != method.getAnnotation(PreDestroy.class)) {
-            method.setAccessible(true);
-            return Collections.<LifecycleAction>singletonList(new LifecycleAction() {
-                @Override
-                public void call(Object obj) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
-                    method.invoke(obj);
-                }
-                
+	public List<LifecycleAction> getActionsForType(final Class<?> type) {
+    	final List<LifecycleAction> typeActions;
+		if (AutoCloseable.class.isAssignableFrom(type)) {
+			LifecycleAction closeableAction = new LifecycleAction() {				
+				@Override
+				public void call(Object obj) throws Exception {
+					((AutoCloseable)obj).close();					
+				}
+				
                 @Override
                 public String toString() {
                     return new StringBuilder()
                         .append("PreDestroy[")
                         .append(type.getName())
                         .append("#")
-                        .append(method.getName())
+                        .append("close")
                         .append("]")
                         .toString();
-                }
-            });
-        }
-        return Collections.emptyList();
+                }				
+			};
+			typeActions = Collections.singletonList(closeableAction);
+		}
+		else {
+	        typeActions = super.getActionsForType(type);
+		}
+		return typeActions;
+	}
+
+	@Override
+    protected List<LifecycleAction> getMethodActions(Set<String> typeContext, final Class<?> type, final Method method) {
+		List<LifecycleAction> typeActions=null;
+    	int modifiers = method.getModifiers();
+		if (!Modifier.isStatic(modifiers) && !Modifier.isAbstract(modifiers) && method.getParameterCount() == 0  && Void.TYPE.equals(method.getReturnType())) {
```

Comment:
We should probably throw an exception if someone is annotating a static method with @PreDestroy or @PostConstruct as it won't work as they would be expecting.